### PR TITLE
ST6RI-547: Render members of State Space Representation (SSR) in Action

### PR DESCRIPTION
**Pull request by himi**
Actions of SSR have `input`, `output`, and `stateSpace` but they were not rendered properly.  `input` and `output` are features having FeatureDirectionKind so that the visualizer could render these features but it did not render their vector components.  `stateSpace` does not have FeatureDirectionKind and it was not rendered at all.

I also added two SSR examples, CartSample and EVSample, in `sysml/src/examples/State Space Representation Examples` directory.

### CODE REVIEW

**Comment by himi** [comment]
This fix conservatively renders these elements in Action, in the sense that it does not cover any kind of VectorQuantityValue because so far only SSR uses VectorQuantityValue .  But it can be extended.

**Comment by himi** [comment]
By this fix, CartSample is rendered with `SHOWINHERITED` style as
Screen Shot 2022-06-02 at 10 29 36 PM
:

File: sysml/src/examples/State Space Representation Examples/CartSample.sysml [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+    import StateSpaceRepresentation::*;
+    part def Cart {
+        attribute mass :> ISQ::mass;
+
+        attribute def CartInput :> Input {
+            attribute force :> ISQ::force;
+        }
+
+        attribute def CartOutput :> Output {
+            attribute velocity :> ISQ::speed;
```
Comment (hpdekoning) [inline sysml/src/examples/State Space Representation Examples/CartSample.sysml:13]: Minor comment is that according to ISO 80000-3 velocity is used for the vector quantity and speed for the scalar quantity, i.e. the name for the magnitude of the velocity, independent of direction. You may want to use `speed` in these examples.

File: sysml/src/examples/State Space Representation Examples/EVSample.sysml [hunk 2]
```diff
@@ -0,0 +1,313 @@
+// State Space Representation EV example
+package EVSample {
+    import SI::*;
+    import StateSpaceRepresentation::*;
+
+    attribute <'A⋅h'> 'ampere hour'  : ElectricChargeUnit = A*h;
```
Comment (hpdekoning) [inline sysml/src/examples/State Space Representation Examples/EVSample.sysml:6]: I will add 'ampere hour' to the TODO list of units to be added into the SI library.

**Review by hpdekoning — APPROVED** [review]

File: sysml/src/examples/State Space Representation Examples/CartSample.sysml (hunk 1 above)
Comment (himi) [reply sysml/src/examples/State Space Representation Examples/CartSample.sysml:13]: Thank you.  I did not know such difference.  This is 1D example and as you said, `speed` is relevant.  I used velocity in EV example but it subsets `ISQ::speed` as well.  Maybe we should have named it to speed.

File: sysml/src/examples/State Space Representation Examples/EVSample.sysml (hunk 2 above)
Comment (himi) [reply sysml/src/examples/State Space Representation Examples/EVSample.sysml:6]: Thank you.  After that I will remove that.

**Comment by seidewitz** [comment]
I cannot reproduce the diagram shown above when visualizing in Eclipse. Visualizing the `context` part in an action diagram with the "Show elements of standard libraries" and "Show inherited members" options set produces:
image

**Comment by seidewitz** [comment]
Trying to open `EVSample` caused a StackOverFlowError:
```
java.lang.StackOverflowError
	at org.eclipse.emf.common.util.BasicEList.contains(BasicEList.java:152)
	at org.eclipse.emf.common.util.AbstractEList.add(AbstractEList.java:298)
	at org.eclipse.uml2.common.util.DerivedEObjectEList$DerivedListIterator.scanNext(DerivedEObjectEList.java:82)
	at org.eclipse.uml2.common.util.DerivedEObjectEList$DerivedListIterator.prepareNext(DerivedEObjectEList.java:96)
	at org.eclipse.uml2.common.util.DerivedEObjectEList$DerivedListIterator.hasNext(DerivedEObjectEList.java:152)
	at org.eclipse.uml2.common.util.DerivedEObjectEList$DerivedListIterator.next(DerivedEObjectEList.java:161)
	at java.base/java.util.Iterator.forEachRemaining(Iterator.java:133)
	at java.base/java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1801)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
	at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:150)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:173)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.forEachOrdered(ReferencePipeline.java:502)
	at org.omg.sysml.lang.sysml.impl.TypeImpl.getOwnedFeature(TypeImpl.java:235)
	at org.omg.sysml.lang.sysml.impl.ExpressionImpl.getInput(ExpressionImpl.java:161)
	at org.omg.sysml.lang.sysml.impl.TypeImpl.directionOf(TypeImpl.java:645)
	at org.omg.sysml.lang.sysml.impl.FeatureImpl.directionFor(FeatureImpl.java:787)
	at org.omg.sysml.util.FeatureUtil.getDirection(FeatureUtil.java:100)
	at org.omg.sysml.util.FeatureUtil.isInputParameter(FeatureUtil.java:86)
	at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:176)
	at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:177)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
	at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:177)
	at java.base/java.util.Iterator.forEachRemaining(Iterator.java:133)
	at java.base/java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1801)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
	at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:150)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:173)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.forEachOrdered(ReferencePipeline.java:502)
	at org.omg.sysml.lang.sysml.impl.InvocationExpressionImpl.getArgument(InvocationExpressionImpl.java:125)
	at org.omg.sysml.lang.sysml.impl.OperatorExpressionImpl$OperandEList.delegateList(OperatorExpressionImpl.java:247)
	at org.eclipse.emf.common.util.DelegatingEList.delegateIsEmpty(DelegatingEList.java:86)
	at org.eclipse.emf.common.util.DelegatingEList.isEmpty(DelegatingEList.java:77)
	at org.omg.sysml.plantuml.VPath$PCFeatureChainExpression.getTargetExp(VPath.java:154)
	at org.omg.sysml.plantuml.VPath$PCFeatureChainExpression.getTarget(VPath.java:163)
	at org.omg.sysml.plantuml.VPath$PC.match(VPath.java:113)
	at org.omg.sysml.plantuml.VPath$PC.enter(VPath.java:127)
	at org.omg.sysml.plantuml.VPath$RefPC.enter(VPath.java:396)
	at org.omg.sysml.plantuml.VPath.enter(VPath.java:541)
	at org.omg.sysml.plantuml.VTraverser.traverse(VTraverser.java:128)
	at org.omg.sysml.plantuml.VTraverser.traverse(VTraverser.java:147)
	at org.omg.sysml.plantuml.VActionMembers.startAction(VActionMembers.java:73)
	at org.omg.sysml.plantuml.VAction.addAction(VAction.java:61)
	at org.omg.sysml.plantuml.VAction.caseActionUsage(VAction.java:73)
	at org.omg.sysml.plantuml.VAction.caseActionUsage(VAction.java:1)
	at org.omg.sysml.lang.sysml.util.SysMLSwitch.doSwitch(SysMLSwitch.java:1143)
	at org.eclipse.emf.ecore.util.Switch.doSwitch(Switch.java:53)
	at org.eclipse.emf.ecore.util.Switch.doSwitch(Switch.java:69)
	at org.omg.sysml.plantuml.Visitor.visit(Visitor.java:656)
	at org.omg.sysml.plantuml.VActionMembers.caseActionUsage(VActionMembers.java:109)
	at org.omg.sysml.plantuml.VActionMembers.caseActionUsage(VActionMembers.java:1)
	at org.omg.sysml.lang.sysml.util.SysMLSwitch.doSwitch(SysMLSwitch.java:1143)
	at org.eclipse.emf.ecore.util.Switch.doSwitch(Switch.java:53)
	at org.eclipse.emf.ecore.util.Switch.doSwitch(Switch.java:69)
	at org.omg.sysml.plantuml.Visitor.visit(Visitor.java:656)
	at org.omg.sysml.plantuml.VTraverser.visitMembership(VTraverser.java:152)
	at org.omg.sysml.plantuml.VTraverser.caseMembership(VTraverser.java:157)
	at org.omg.sysml.plantuml.VTraverser.caseMembership(VTraverser.java:1)
	at org.omg.sysml.lang.sysml.util.SysMLSwitch.doSwitch(SysMLSwitch.java:397)
	at org.eclipse.emf.ecore.util.Switch.doSwitch(Switch.java:53)
	at org.eclipse.emf.ecore.util.Switch.doSwitch(Switch.java:69)
	at org.omg.sysml.plantuml.Visitor.visit(Visitor.java:656)
	at org.omg.sysml.plantuml.VTraverser.traverseInherited(VTraverser.java:113)
	at org.omg.sysml.plantuml.VTraverser.traverse(VTraverser.java:135)
	at org.omg.sysml.plantuml.VTraverser.traverse(VTraverser.java:147)
	at org.omg.sysml.plantuml.VActionMembers.startAction(VActionMembers.java:73)
	at org.omg.sysml.plantuml.VAction.addAction(VAction.java:61)
	at org.omg.sysml.plantuml.VAction.caseActionUsage(VAction.java:73)
	at org.omg.sysml.plantuml.VAction.caseActionUsage(VAction.java:1)
	at org.omg.sysml.lang.sysml.util.SysMLSwitch.doSwitch(SysMLSwitch.java:1143)
	at org.eclipse.emf.ecore.util.Switch.doSwitch(Switch.java:53)
	at org.eclipse.emf.ecore.util.Switch.doSwitch(Switch.java:69)
	at org.omg.sysml.plantuml.Visitor.visit(Visitor.java:656)
	at org.omg.sysml.plantuml.VActionMembers.caseActionUsage(VActionMembers.java:109)
	at org.omg.sysml.plantuml.VActionMembers.caseActionUsage(VActionMembers.java:1)
	at org.omg.sysml.lang.sysml.util.SysMLSwitch.doSwitch(SysMLSwitch.java:1143)
	at org.eclipse.emf.ecore.util.Switch.doSwitch(Switch.java:53)
	at org.eclipse.emf.ecore.util.Switch.doSwitch(Switch.java:69)
	at org.omg.sysml.plantuml.Visitor.visit(Visitor.java:656)
	at org.omg.sysml.plantuml.VTraverser.visitMembership(VTraverser.java:152)
	at org.omg.sysml.plantuml.VTraverser.caseMembership(VTraverser.java:157)
	at org.omg.sysml.plantuml.VTraverser.caseMembership(VTraverser.java:1)
	at org.omg.sysml.lang.sysml.util.SysMLSwitch.doSwitch(SysMLSwitch.java:397)
	at org.eclipse.emf.ecore.util.Switch.doSwitch(Switch.java:53)
	at org.eclipse.emf.ecore.util.Switch.doSwitch(Switch.java:69)
        ...
```

**Comment by himi** [comment]
@seidewitz,  Hmm, in ST6RI-547 branch, I could not reproduce both cases you showed...  I'll check the environment again but did you merge or change something in your environment?

**Comment by seidewitz** [comment]
I did a pull and ran on the ST6RI-547 branch.

**Comment by seidewitz** [comment]
I did a Maven clean package rebuild, and it works now.

**Review by seidewitz — APPROVED** [review]

**Pull request merged**